draft-ietf-masque-h3-datagram.md

```diff
@@ -181,74 +181,82 @@ all cases, the maximum permitted value of the H3_DATAGRAM SETTINGS parameter is
 1.
 
 
-# Datagram-Flow-Id Header Field Definition {#header}
+# REGISTER_DATAGRAM_FLOW_ID HTTP/3 Frame Definition {#register-frame}
 
-"Datagram-Flow-Id" is a List Structured
-Field {{!STRUCT-FIELD=I-D.ietf-httpbis-header-structure}}, whose members MUST
-all be Items of type Integer. Integers MUST be non-negative. Its ABNF is:
+The REGISTER_DATAGRAM_FLOW_ID frame (type=TBD) is used to register a flow ID
+with the QUIC stream that it is sent on. By sending this frame, the sender
+indicates that it will process any received datagram with the corresponding
+flow ID using semantics specific to this frame. For example, if this is sent on
+a client-initiated bidirectional stream that has carried an HTTP request, this
+flow ID will now be interpreted as governed by the method of that request.
 
 ~~~
-  Datagram-Flow-Id = sf-list
+REGISTER_DATAGRAM_FLOW_ID Frame {
+  Type (i) = TBD,
+  Length (i),
+  Flow ID (i),
+  Encoded Field Section (..),
+}
 ~~~
+{: #register-frame-format title="REGISTER_DATAGRAM_FLOW_ID HTTP/3 Frame Format"}
 
-The "Datagram-Flow-Id" header field is used to associate one or more datagram
-flows with an HTTP message. As a simple example using a single
-flow, the definition of an HTTP method could instruct the client to use
-its flow ID allocation service to allocate a new flow ID, and
-then the client will add the "Datagram-Flow-Id" header field to its request to
-communicate that value to the server. In this example, the resulting header
-field could look like:
+The Type and Length fields follows the definition of HTTP/3 frames from {{H3}}.
+The payload consists of:
 
-~~~
-  Datagram-Flow-Id = 2
-~~~
+Flow ID:
 
-List members are flow ID elements, which can be named or unnamed.
-One element in the list is allowed to be unnamed, but all but one elements
-MUST carry a name. The name of an element is encoded in the key of the first
-parameter of that element (parameters are defined in Section 3.1.2 of
-{{STRUCT-FIELD}}). Each name MUST NOT appear more than once in the list. The
-value of the first parameter of each named element (whose corresponding key
-conveys the element name) MUST be of type Boolean and equal to true. The value
-of the first parameter of the unnamed element MUST NOT be of type Boolean. The
-ordering of the list does not carry any semantics. For example, an HTTP method
-that wishes to use four datagram flows for the lifetime of its
-request stream could look like this:
+: The flow ID to associate with the QUIC stream that carried this frame.
 
-~~~
-  Datagram-Flow-Id = 42, 44; ecn-ect0, 46; ecn-ect1, 48; ecn-ce
-~~~
+Encoded Field Section:
+
+: QPACK-encoded fields. This allows extensions to optionally attach headers to
```

Comment:
I think this needs some more explanation; what use cases motivate this extensibility?

Unless there is a good reason to use QPACK I wouldn't. QPACK is for HTTP message metadata and these are clearly not those. I can predict this to cause all sorts of oddites/attacks when people try to reconstruct this field section and accidentally combine it with one reconstructed from HEADERS - or just the simple case of blocking/poisoning the QPACK dynamic table.

Could we satisfy the extensibility requirements with a simple value-only field of `sf-list`?

Reply:
Agreed, it'd be good to understand what our extensibility requirements are here.  I can imagine a fairly wide range.

Reply:
I agree we should be looking at what the extensibility requirements are, and providing the appropriate degree of freedom here.

I disagree that QPACK should only be used for HTTP message metadata and disallowed in extensions.  Why are we worried that this field section would be combined incorrectly with the header field section, but we're not worried about that for trailers?  Similar question for blocking/poisoning the QPACK table?

Reply:
I don't know what these frames are in relation to HTTP semantics. They contain an encoded field section but to where does that field section apply? There are rules and documented limitiations for handling trailers https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#trailers.limitations.

This readslike a weird form of "middlers".

Mixing the use of a compression context between these HEADERS and extensions might be ok, or it might be horribly insecure. Is there precidence for extensions that have done this before?

Maybe extensions can use QPACK but I want to know if H3_DATAGRAM needs to eat that complexity.

Reply:
Based on this discussion, QPACK isn't as simple as I originally thought, so I replaced it with a simple custom encoding.

Reply:
Hm, I think QPACK ended up perhaps too tightly coupled to HTTP semantics - at least it references Field Lines and Field Sections.  The specification itself is otherwise devoid of semantic meaning - it's just an encoding and compression scheme for name/value pairs.  I think it would be kind of a bummer if it couldn't be used, particularly by HTTP extensions, to convey compressed, structured metadata - like "middlers", etc.

But I can see @DavidSchinazi giving us a nudge to quit discussing QPACK here...

Reply:
Oh I'm in complete agreement that HPACK/QPACK are very nice generic key-value compression schemes. It's slightly sad they are entwined with messages but that's what we've been dealt.